File: src/ObjReader.java
import java.io.*;
import java.util.ArrayList;

/**
 * Created by denak on 13.12.2017.
 */
public class ObjReader {

    public static void save(Obj obj, String filename) {
        StringBuilder points = new StringBuilder();
        StringBuilder normals = new StringBuilder();
        StringBuilder textures = new StringBuilder();
        StringBuilder triangles = new StringBuilder();

        for (int i = 0, t = 1, n = 1; i < obj.triangles.size(); i++) {
            Triangle triangle = obj.triangles.get(i);
            String a = "v " + String.valueOf(triangle.a.x) + " " + String.valueOf(triangle.a.y) + " " + String.valueOf(triangle.a.z) + "\n";
            String b = "v " + String.valueOf(triangle.b.x) + " " + String.valueOf(triangle.b.y) + " " + String.valueOf(triangle.b.z) + "\n";
            String c = "v " + String.valueOf(triangle.c.x) + " " + String.valueOf(triangle.c.y) + " " + String.valueOf(triangle.c.z) + "\n";
            String na = "", nb = "", nc = "";
            String ta = "", tb = "", tc = "";
            boolean is_textures = false;
            boolean is_normals = false;
            if (triangle.normal_a != null) {
                is_normals = true;
                na = "vn " + String.valueOf(triangle.normal_a.x) + " " + String.valueOf(triangle.normal_a.y) + " " + String.valueOf(triangle.normal_a.z) + "\n";
                n++;
            }
            if (triangle.normal_b != null) {
                nb = "vn " + String.valueOf(triangle.normal_b.x) + " " + String.valueOf(triangle.normal_b.y) + " " + String.valueOf(triangle.normal_b.z) + "\n";
                n++;
            }
            if (triangle.normal_c != null) {
                nc = "vn " + String.valueOf(triangle.normal_c.x) + " " + String.valueOf(triangle.normal_c.y) + " " + String.valueOf(triangle.normal_c.z) + "\n";
                n++;
            }
            if (triangle.texture_a != null) {
                is_textures = true;
                ta = "vt " + String.valueOf(triangle.texture_a.u) + " " + String.valueOf(triangle.texture_a.v) + "\n";
                t++;
            }
            if (triangle.texture_b != null) {
                tb = "vt " + String.valueOf(triangle.texture_b.u) + " " + String.valueOf(triangle.texture_b.v) + "\n";
                t++;
            }
            if (triangle.texture_c != null) {
                tc = "vt " + String.valueOf(triangle.texture_c.u) + " " + String.valueOf(triangle.texture_c.v) + "\n";
                t++;
            }
            String tr;
            if (is_normals) {
                if (is_textures) {
                    tr = "f " + (i * 3 + 1) + "/" + (t - 2) + "/" + (n - 2) + " " + (i * 3 + 2) + "/" + (t - 1) + "/" + (n - 1) + " " + (i * 3 + 3) + "/" + t + "/" + n + "\n";
                } else {
                    tr = "f " + (i * 3 + 1) + "//" + (n - 2) + " " + (i * 3 + 2) + "//" + (n - 1) + " " + (i * 3 + 3) + "//" + n + "\n";
                }
            } else {
                if (is_textures) {
                    tr = "f " + (i * 3 + 1) + "/" + (t - 2) + " " + (i * 3 + 2) + "/" + (t - 1) + " " + (i * 3 + 3) + "/" + t + "\n";
                } else {
                    tr = "f " + (i * 3 + 1) + " " + (i * 3 + 2) + " " + (i * 3 + 3) + "\n";
                }
            }
            points.append(a).append(b).append(c);
            textures.append(ta).append(tb).append(tc);
            normals.append(na).append(nb).append(nc);
            triangles.append(tr);
        }
        try (FileWriter writer = new FileWriter(filename)) {
            writer.write(points.toString());
            writer.write(textures.toString());
            writer.write(normals.toString());
            writer.write(triangles.toString());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static Obj load(String filename) {
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            ArrayList<Point3D> points = new ArrayList<>();
            ArrayList<Point3D> normals = new ArrayList<>();
            ArrayList<Point2D> textures = new ArrayList<>();
            ArrayList<Triangle> triangles = new ArrayList<>();

            while (reader.ready()) {
                String line = reader.readLine();
                String elements[] = line.split(" ");
                if (elements.length == 4) {
                    elements[1] = elements[2];
                    elements[2] = elements[3];
                }
                if (elements.length == 5) {
                    elements[1] = elements[3];
                    elements[2] = elements[4];
                }
                switch (elements[0]) {
                    case "v": {
                        float x = Float.parseFloat(elements[1]);
                        float y = Float.parseFloat(elements[2]);
                        float z = Float.parseFloat(elements[3]);

                        points.add(new Point3D(x, y, z));
                        break;
                    }
                    case "vn": {
                        float x = Float.parseFloat(elements[1]);
                        float y = Float.parseFloat(elements[2]);
                        float z = Float.parseFloat(elements[3]);

                        normals.add(new Point3D(x, y, z));
                        break;
                    }
                    case "vt": {
                        float u = Float.parseFloat(elements[1]);
                        float v = Float.parseFloat(elements[2]);

                        textures.add(new Point2D(u, v));
                        break;
                    }
                    case "f": {
                        Point3D fPoints[] = new Point3D[3];
                        Point3D fNormals[] = new Point3D[3];
                        Point2D fTextures[] = new Point2D[3];

                        for (int i = 0; i < 3; i++) {
                            String[] coords = elements[i + 1].split("/");
                            fPoints[i] = points.get(Integer.parseInt(coords[0]) - 1);
                            switch (coords.length) {
                                case 3: {
                                    fNormals[i] = normals.get(Integer.parseInt(coords[2]) - 1);
                                    if (coords[1].equals("")) {
                                        break;
                                    }
                                }
                                case 2: {
                                    fTextures[i] = textures.get(Integer.parseInt(coords[1]) - 1);
                                }
                            }
                        }
                        Triangle triangle = new Triangle(
                                fPoints[0], fPoints[1], fPoints[2],
                                fNormals[0], fNormals[1], fNormals[2],
                                fTextures[0], fTextures[1], fTextures[2]);
                        triangles.add(triangle);
                    }
                }
            }
            return new Obj(triangles);
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: src/ObjReader.kt
import java.io.*

/**
 * Created by denak on 13.12.2017.
 */
object ObjReader {

    /*public static void save(Obj obj, String filename) {
        ArrayList<Point3D> points = new ArrayList<>();
        ArrayList<Point3D> normals = new ArrayList<>();
        ArrayList<Point2D> textures = new ArrayList<>();

        for (Triangle triangle : obj.triangles) {
            if (!points.contains(triangle.getA())) {
                points.add(triangle.getA());
            }
            if (!points.contains(triangle.getB())) {
                points.add(triangle.getB());
            }
            if (!points.contains(triangle.getC())) {
                points.add(triangle.getC());
            }
            if (triangle.hasNormalCoords()) {
                if (!normals.contains(triangle.getNormalA())) {
                    normals.add(triangle.getNormalA());
                }
                if (!normals.contains(triangle.getNormalB())) {
                    normals.add(triangle.getNormalB());
                }
                if (!normals.contains(triangle.getNormalC())) {
                    normals.add(triangle.getNormalC());
                }
            }
            if (triangle.hasTextureCoords()) {
                if (!textures.contains(triangle.getTextureA())) {
                    textures.add(triangle.getTextureA());
                }
                if (!textures.contains(triangle.getTextureB())) {
                    textures.add(triangle.getTextureB());
                }
                if (!textures.contains(triangle.getTextureC())) {
                    textures.add(triangle.getTextureC());
                }
            }
        }
        points.sort(Comparator.comparingInt(o -> o.getId()));
        normals.sort(Comparator.comparingInt(o -> o.getId()));
        textures.sort(Comparator.comparingInt(o -> o.getId()));
        StringBuilder string_points = new StringBuilder();
        StringBuilder string_normals = new StringBuilder();
        StringBuilder string_textures = new StringBuilder();
        StringBuilder string_triangles = new StringBuilder();

        for (Point3D point : points) {
            String p = "v " + String.valueOf(point.getX()) + " " +
                    String.valueOf(point.getY()) + " " +
                    String.valueOf(point.getZ()) + "\n";
            string_points.append(p);
        }

        for (Point3D point : normals) {
            String n = "vn " + String.valueOf(point.getX()) + " " +
                    String.valueOf(point.getY()) + " " +
                    String.valueOf(point.getZ()) + "\n";
            string_normals.append(n);
        }

        for (Point2D point : textures) {
            String t = "vt " + String.valueOf(point.getU()) + " " +
                    String.valueOf(point.getV()) + "\n";
            string_textures.append(t);
        }
        for (Triangle triangle : obj.triangles) {
            String tr;
            if (triangle.hasNormalCoords()) {
                if (triangle.hasTextureCoords()) {
                    tr = "f " + triangle.getA().getId() + "/" + triangle.getTextureA().getId() + "/" + triangle.getNormalA().getId() + " " +
                            triangle.getB().getId() + "/" + triangle.getTextureB().getId() + "/" + triangle.getNormalB().getId() + " " +
                            triangle.getC().getId() + "/" + triangle.getTextureC().getId() + "/" + triangle.getNormalC().getId() + "\n";
                } else {
                    tr = "f " + triangle.getA().getId() + "//" + triangle.getNormalA().getId() + " " +
                            triangle.getB().getId() + "//" + triangle.getNormalB().getId() + " " +
                            triangle.getC().getId() + "//" + triangle.getNormalC().getId() + "\n";
                }
            } else {
                if (triangle.hasTextureCoords()) {
                    tr = "f " + triangle.getA().getId() + "/" + triangle.getTextureA().getId() + " " +
                            triangle.getB().getId() + "/" + triangle.getTextureB().getId() + " " +
                            triangle.getC().getId() + "/" + triangle.getTextureC().getId() + "\n";
                } else {
                    tr = "f " + triangle.getA().getId() + " " +
                            triangle.getB().getId() + " " +
                            triangle.getC().getId() + "\n";
                }
            }
            string_triangles.append(tr);
        }

        try (FileWriter writer = new FileWriter(filename)) {
            //writer.write("# points\n");
            writer.write(string_points.toString());
            //writer.write("# textures\n");
            writer.write(string_textures.toString());
            //writer.write("# normals\n");
            writer.write(string_normals.toString());
            //writer.write("# triangles\n");
            writer.write(string_triangles.toString());
            writer.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }*/

    fun load(filename: String): Obj? {
        try {
            BufferedReader(FileReader(filename)).use { reader ->
                val points = ArrayList<Point3D>()
                val normals = ArrayList<Point3D>()
                val textures = ArrayList<Point2D>()
                val triangles = ArrayList<Triangle>()
                var p = 1
                var n = 1
                var t = 1

                while (reader.ready()) {
                    var line = reader.readLine()
                    line = line.replace(" {2}".toRegex(), " ")
                    val elements = line.split(" ".toRegex())
                    when (elements[0]) {
                        "v" -> {
                            val x = elements[1].toFloat()
                            val y = elements[2].toFloat()
                            val z = elements[3].toFloat()

                            points.add(Point3D(x, y, z, p))
                            p++
                        }
                        "vn" -> {
                            val x = elements[1].toFloat()
                            val y = elements[2].toFloat()
                            val z = elements[3].toFloat()

                            normals.add(Point3D(x, y, z, n))
                            n++
                        }
                        "vt" -> {
                            val u = elements[1].toFloat()
                            val v = elements[2].toFloat()

                            textures.add(Point2D(u, v, t))
                            t++
                        }
                        "f" -> {
                            val fPoints = ArrayList<Point3D>()
                            val fNormals = ArrayList<Point3D>()
                            val fTextures = ArrayList<Point2D>()
                            for (i in 0..2) {
                                val coords = elements[i + 1].split("/".toRegex())
                                fPoints += points[coords[0].toInt() - 1]
                                coords.size
                                when (coords.size) {
                                    3 -> {
                                        fNormals += normals[coords[2].toInt() - 1]
                                        if(coords[1].isNotEmpty()) fTextures += textures[coords[1].toInt() - 1]
                                    }
                                    2 -> fTextures += textures[coords[1].toInt() - 1]

                                }
                            }

                            triangles += if (fNormals.size != 0) if (fTextures.size != 0) Triangle(
                                    fPoints[0], fPoints[1], fPoints[2],
                                    fNormals[0], fNormals[1], fNormals[2],
                                    fTextures[0], fTextures[1], fTextures[2])
                            else Triangle(
                                    fPoints[0], fPoints[1], fPoints[2],
                                    fNormals[0], fNormals[1], fNormals[2])
                            else if (fTextures.size != 0) Triangle(
                                    fPoints[0], fPoints[1], fPoints[2],
                                    fTextures[0], fTextures[1], fTextures[2])
                            else Triangle(fPoints[0], fPoints[1], fPoints[2])


                        }
                    }
                }
                return Obj(triangles)
            }
        } catch (e: IOException) {
            e.printStackTrace()
            return null
        }

    }
}
